### PR TITLE
[tosa] Legalize tfl.atan2 to tosa.table operation

### PR DESCRIPTION
**Pull request by lhutton1**
In a similar fashion to tfl.sin and tlf.cos legalization, a tosa.table provides an atan approximation. Additional logic is then used to determine the correct quadrant of the atan2 function.

Signed-off-by: Luke Hutton <luke.hutton@arm.com>

### CODE REVIEW

**Review by jpienaar — COMMENTED** [review]
Forgot to hit send on comments ...

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 1]
```diff
@@ -2933,6 +2934,145 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  Value input_y = tfl_atan2_op.getY();
+  RankedTensorType input_y_ty = input_y.getType().dyn_cast<RankedTensorType>();
+  Value input_x = tfl_atan2_op.getX();
+  RankedTensorType input_x_ty = input_x.getType().dyn_cast<RankedTensorType>();
+  ShapedType output_ty =
+      tfl_atan2_op.getResult().getType().dyn_cast<ShapedType>();
+
+  if (!input_y_ty || !input_x_ty || !output_ty) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLAtan2Op: input/output types not specified");
```
[the remark this answers is not in the document]
Comment (lhutton1) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:2950]: Thanks for the explanation, will do. In this case does it make sense to also remove the `output_ty` check?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 2]
```diff
@@ -2933,6 +2934,145 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  auto int_scaled =
+      CreateOpAndInfer<tosa::MulOp>(rewriter, loc, input_y_ty, translate,
+                                    int_limit, rewriter.getI32IntegerAttr(0));
+
+  auto int16_ty = input_y_ty.clone(rewriter.getIntegerType(16));
+  auto casted =
+      CreateOpAndInfer<tosa::CastOp>(rewriter, loc, int16_ty, int_scaled);
+
+  // 3. Compute a lookup table using the domain of [0, 1] for atan.
+  auto atan_func = [](double x) -> double { return std::atan(x); };
```
[the remark this answers is not in the document]
Comment (lhutton1) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3027]: Good point, I think I should have mentioned in the PR description that this is intended as a workaround for now (similar to what's mentioned in https://github.com/tensorflow/tensorflow/commit/6df59861af1c95ccfe846456b0eecbdcafa7455f). Perhaps leaving a similar note as a comment (in code) might help?

**Review by lhutton1 — COMMENTED** [review]
Thanks for the review @jpienaar! I left some responses below

**Comment by gbaned** [comment]
Hi @lhutton1 Can you please resolve conflicts? Thank you!

**Comment by lhutton1** [comment]
Friendly ping @jpienaar

File: tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir [hunk 3]
```diff
@@ -640,6 +640,47 @@ func.func @test_cos(%arg0: tensor<10xf32>) -> tensor<*xf32> {
 
 // -----
 
+// CHECK-LABEL: test_atan2
+// CHECK-SAME: -> tensor<13x21x3xf32>
+// CHECK: %[[VAL_2:.*]] = "tosa.const"() {value = dense<2.000000e+00> : tensor<1x1x1xf32>} : () -> tensor<1x1x1xf32>
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir:645]: Prefer CHECK-DAG for constants (there is an internal check for this that ends up blocking presubmits, but seems not upstream)

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 4]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+      tfl_atan2_op.getResult().getType().dyn_cast<ShapedType>();
+
+  if (!input_y_ty || !input_x_ty) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLAtan2Op: ranked inputs required");
+  }
+
+  Type input_y_ety = input_y_ty.getElementType();
+  Type input_x_ety = input_x_ty.getElementType();
+  Type output_ety = output_ty.getElementType();
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3020]: Check for output_ty above too to avoid null here or just use cast above - I don't think TFL::Atan2Op is legal unless it produces a ShapedType

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 5]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
   return success();
 }
 
+LogicalResult ConvertTFLAtan2Op::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_atan2_op = cast<TFL::Atan2Op>(op);
+  Location loc = op->getLoc();
+  Value input_y = tfl_atan2_op.getY();
+  RankedTensorType input_y_ty = input_y.getType().dyn_cast<RankedTensorType>();
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3007]: With the new casting mechanisms in LLVM one can now finally do

auto input_y_ty = dyn_cast<RankedTensorType>(input_y.getType());

prefer to not repeat type for these (for those that are outputs of cast's we use auto as type is clear locally).

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 6]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  if (!input_y_ty || !input_x_ty) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLAtan2Op: ranked inputs required");
+  }
+
+  Type input_y_ety = input_y_ty.getElementType();
+  Type input_x_ety = input_x_ty.getElementType();
+  Type output_ety = output_ty.getElementType();
+
+  bool op_is_fp = input_y_ty.getElementType().isF32();
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3022]: Use input_y_ety?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 7]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  Value input_y = tfl_atan2_op.getY();
+  RankedTensorType input_y_ty = input_y.getType().dyn_cast<RankedTensorType>();
+  Value input_x = tfl_atan2_op.getX();
+  RankedTensorType input_x_ty = input_x.getType().dyn_cast<RankedTensorType>();
+  ShapedType output_ty =
+      tfl_atan2_op.getResult().getType().dyn_cast<ShapedType>();
+
+  if (!input_y_ty || !input_x_ty) {
+    return rewriter.notifyMatchFailure(
+        op, "ConvertTFLAtan2Op: ranked inputs required");
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3015]: Update these to keep it consistent with the others (skipping the pattern name in error message)

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 8]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  // arctan(-z) = -arctan(z)                                        (1)
+
+  Value pi = getTosaConstTensorSingleF32(rewriter, op, M_PI);
+  Value pi_2 = getTosaConstTensorSingleF32(rewriter, op, M_PI_2);
+  Value zero = getTosaConstTensorSingleF32(rewriter, op, 0.0);
+  Value one = getTosaConstTensorSingleF32(rewriter, op, 1.0);
+  Value two = getTosaConstTensorSingleF32(rewriter, op, 2.0);
+
+  // 1. Restrict the input to the atan lookup from [-inf, inf] to [0, 1].
+  // By utilizing (0) and (1) we compute: min(|x|, |y|) / max(|x|, |y|).
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3046]: Do we need to handle case where |x| and |y| are both 0?

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 9]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  auto max_xy = CreateOpAndInfer<tosa::MaximumOp>(rewriter, loc, input_y_ty,
+                                                  abs_y, abs_x);
+  auto recip =
+      CreateOpAndInfer<tosa::ReciprocalOp>(rewriter, loc, input_y_ty, max_xy);
+  auto atan_input = CreateOpAndInfer<tosa::MulOp>(
+      rewriter, loc, input_y_ty, recip, min_xy, rewriter.getI32IntegerAttr(0));
+
+  // 2. Scale and translate the normalized domain to the table domain. This
+  // includes a translating and scaling to [-int16_max, int16_max] and casting
+  // to an i16.
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3062]: Add reason for casting here too.

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc [hunk 10]
```diff
@@ -2998,6 +2999,138 @@ LogicalResult ConvertTFLCosOp::matchAndRewrite(
...
+  // 3. Compute a lookup table using the domain of [0, 1] for atan.
+  // Note: the implementation of std::atan2 may be different on
+  // different machines, so may result in varying numerical results.
+  auto atan_func = [](double x) -> double { return std::atan(x); };
+  Value table_const = getTosaConst16bitTable(rewriter, op, atan_func, 0.0, 1.0);
+  auto table_result = CreateOpAndInfer<tosa::TableOp>(
+      rewriter, loc, output_ty.clone(rewriter.getIntegerType(32)), casted,
+      table_const);
+
+  // 4. The range of table is a 23-bit two's compliment value. Normalize the
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:3091]: complement

File: tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc [hunk 11]
```diff
@@ -313,13 +313,9 @@ Value getTosaConst16bitTable(PatternRewriter& rewriter, Operation* op,
...
       static_cast<int16_t>(std::min(std::max(max_val, -32768), 32767)));
 
-  auto element_qtype =
-      UniformQuantizedType::get(true, rewriter.getIntegerType(16),
-                                rewriter.getF32Type(), 1.0f, 0, -32768, 32767);
-  auto const_type = tensorflow::GetTypeFromTFTensorShape({513}, element_qtype);
-  auto storage_type = tensorflow::GetTypeFromTFTensorShape(
-      {513}, element_qtype.getStorageType());
-  auto const_attr = DenseElementsAttr::get(storage_type, llvm::ArrayRef(table));
+  auto const_type = tensorflow::GetTypeFromTFTensorShape({513}, rewriter.getIntegerType(16));
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc:316]: Why was this change needed?

**Review by jpienaar — COMMENTED** [review]
Thanks for the ping

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc (hunk 8 above)
Comment (jpienaar) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:2951]: For some reason I don't see your responses.

**Review by lhutton1 — COMMENTED** [review]
Apologies for the late follow-up on this and thanks for the review @jpienaar, please find my replies below

File: tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc (hunk 11 above)
Comment (jpienaar) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc:316]: If everything still works for this I'm happy, I don't have context as to why this was like this originally.

**Review by jpienaar — APPROVED** [review]
LG, check on consistency part and if the behavior TOSA side in case there were 0's but beyond that good. (and I saw your responses later, don't understand github PRs some days)

**Comment by lhutton1** [comment]
Thanks! It was actually me who forgot to hit send 😅

File: tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc (hunk 8 above)
Comment (eric-k256) [reply tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:2951]: Having consistent behavior is nice, but this is undefined behavior, so it may be impossible to guarantee consistency in all cases.

This sequence of ops does have defined behavior in TOSA, according to the current spec. We don't define reciprocal(0) explicitly, but we do say non-zero / 0 = infinity, and then 0 * infinity = 0.

Based on this, I think this solution looks good.
